src/zjs_spi.c

```diff
@@ -0,0 +1,319 @@
+// Copyright (c) 2017, Intel Corporation.
+
+// Zephyr includes
+#ifndef ZJS_LINUX_BUILD
+#include <zephyr.h>
+#include <spi.h>
+#include <misc/util.h>
+#endif
+#include <string.h>
+#include <stdlib.h>
+
+// ZJS includes
+#include "zjs_spi.h"
+#include "zjs_util.h"
+#include "zjs_buffer.h"
+#include "zjs_callbacks.h"
+#include "zjs_error.h"
+
+#define SPI_BUS "SPI_"
+#define MAX_SPI_BUS 127
+#define MAX_READ_BUFF 1000
+#define MAX_DIR_LEN 13
+
+enum spi_topology {
+    ZJS_TOPOLOGY_FULL_DUPLEX,
+    ZJS_TOPOLOGY_READ,
+    ZJS_TOPOLOGY_WRITE,
+    ZJS_TOPOLOGY_MULTIPLEXED,
+    ZJS_TOPOLOGY_DAISY_CHAIN
+};
+
+typedef struct spi_handle {
+    struct device *spi_device;
+    enum spi_topology topology;
+    bool closed;
+} spi_handle_t;
+
+static jerry_value_t zjs_spi_prototype;
+
+static void zjs_spi_callback_free(void *native)
+{
+    // requires: handle is the native pointer we registered with
+    //             jerry_set_object_native_handle
+    //  effects: frees the spi item
+    spi_handle_t *handle = (spi_handle_t *)native;
+    handle->closed = true;
+    zjs_free(handle);
+}
+
+static const jerry_object_native_info_t spi_type_info =
+{
+   .free_cb = zjs_spi_callback_free
+};
+
+static ZJS_DECL_FUNC(zjs_spi_transceive)
+{
+    // requires: Writes and reads from the SPI bus, takes one argument
+    //           arg[0] - Buffer to write to the SPI bus
+    //  effects: Writes the buffer to the SPI bus and returns the buffer
+    //           received on the SPI in reply.
+
+    ZJS_VALIDATE_ARGS(Z_NUMBER, Z_OPTIONAL Z_ARRAY Z_STRING, Z_OPTIONAL Z_STRING);
+
+    // No more than 3 args
+    if (argc > 3)
+        return ZJS_STD_ERROR(RangeError, "Invalid args");
+
+    ZJS_GET_HANDLE(this, spi_handle_t, handle, spi_type_info);
+
+    if (handle->closed == true) {
+        ZJS_PRINT("SPI bus is closed\n");
+        return jerry_create_null();
+    }
+
+    size_t len;
+    jerry_value_t buffer;
+    zjs_buffer_t *tx_buf;
+    zjs_buffer_t *rx_buf;
+    jerry_value_t rx_buf_obj = jerry_create_null();
+    jerry_value_t tx_buf_obj;
+
+    if (spi_slave_select(handle->spi_device, jerry_get_number_value(argv[0]))) {
+        return zjs_error("SPI slave select failed\n");
+    }
+
+    // If only a slave is given, this must be a single read or its invalid
+    if (argc == 1) {
+        if (handle->topology != ZJS_TOPOLOGY_READ) {
+            return ZJS_STD_ERROR(RangeError, "Missing transmit buffer");
+        }
+    }
+
+    char dir_str[13];
+    // Set the direction default based on the topology.
+    enum spi_topology dir_arg = handle->topology;
+
+    // If we have a 'direction' arg, get it and validate
+    if (argc == 3) {
```

Comment:
1) It's a squishy language, and 2) It's good enough for C. :)

I forget, I've done some searching for common practice on the topic. Just now I opened Chrome inspector and typed Date() with a whole bunch of arguments and it was fine with it, just ignored them after the first three.

But if we decide to do it, it makes sense to do it uniformly everywhere using macros rather than repeating the logic in every function. So I took the occasional appearance of such logic out when I added ZJS_VALIDATE_ARGS.